Create stop start processes

chameleon07 opened the pull request:
-added RoboProcesses class
-created process_list dictionary
-did createProcess function

chameleon07 committed:
Modified createProcesses and startAllProcesses

chameleon07 committed:
Created loops for startAllProcesses and stopAllProcesses

chameleon07 committed:
Redo of createProcesses

chameleon07 committed:
Fixed placement of class..

chameleon07 committed:
Fixed placement of list

chameleon07 committed:
loops to go through process_list created for startAllProcesses and stopAllProcesses

ottopasuuna commented on ProcessManager.py:
```diff
@@ -6,14 +6,24 @@
+                Popen(self.set)
+
+
+process_list = {}
```
`process_list` should be inside the ProcessManager class. I was thinking it would replace `self.T`. Also consider naming it something other than "process_list" as its not a list.

chameleon07 replied:
True true

ottopasuuna commented on ProcessManager.py:
```diff
@@ -6,14 +6,24 @@
+            def _init_(self, cmd):
+                self.cmd = cmd
+                self.set = None
+                Popen(self.set)
```
The line `Popen(self.set)` tries to start a program `None` which won't work. This line could probably just be removed.

chameleon07 replied:
Yeah, will do... Wasn't sure if I had to have Popen there or how I am suppose to set it to None... I guess just have Popen executed in startAllProcesses?

ottopasuuna commented on ProcessManager.py:
```diff
@@ -6,14 +6,24 @@
+class RoboProcesses:
+            def _init_(self, cmd):
+                self.cmd = cmd
+                self.set = None
```
What exactly is `set` storing?

ottopasuuna commented on ProcessManager.py:
```diff
@@ -6,14 +6,24 @@
+from subprocess import Popen
+
+class RoboProcesses:
+            def _init_(self, cmd):
```
Should be two underscores for the constructor method (`__init__`). 😃
Also, super nitpicky thing, but the block starting on line 12 should only be intented 4 spaces in keeping with [PEP 8](https://www.python.org/dev/peps/pep-0008/#indentation).

ottopasuuna commented on ProcessManager.py:
```diff
@@ -25,12 +35,17 @@ def isEmpty(self):
         """
 
+        process_list[name] = RoboProcesses(command)
+        return process_list
```
Once `process_list` is an atribute of the ProcessManager, it won't need to be returned here. We could have it so `createProcess` returns true or false if a failure occured.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -42,11 +57,20 @@ def startAllProcesses(self):
+
+        for process in process_list:
+            process.kill()
+            process.wait()
```
In this for loop, `process` is a `RoboProcess` which doesn't have methods `kill()` or `wait()`. If `RoboProcess` stores a `Popen` object, you'll have to access that and call its `kill()` and `wait()` methods.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -42,11 +57,20 @@ def startAllProcesses(self):
         """
 
+        for process in process_list:
+            process_list[process].set = process_list[process].cmd
```
I think this is where the `Popen` constructor should be called.

ottopasuuna replied:
Another thing you could do if you wanted is create another method in `ProcessManager` to start just a single process, and have `startAllProcesses` call that new method for every process.

ottopasuuna reviewed (changes requested):
Awesome start! Just a few things to change, you can reply to my comments if you have questions about them.

chameleon07 committed:
Revisions made

ottopasuuna commented on ProcessManager.py:
```diff
@@ -9,20 +9,21 @@
+    def __init__(self, cmd):
+        self.cmd = cmd
+    def execute(self):
+        Popen(self.cmd)
```
I like this, but it doesn't save the Popen object, so we can't do anything with the process after it's started. We'll have to save the return value of the Popen constructor in `self.popen` or something like that. The `self.popen` atribute should also be initialized to `None` in the `__init__` method to signify that the newly created RoboProcess has not been started yet.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -9,20 +9,21 @@
 
     This class represents the pocessing module for the robocluster operating system.
     """
+process_dict = {}
```
This is no longer global, but it is a class atribute which is the same for every instance of the class, ie if there are two ProcecssManagers and one of them modifies its `process_dict`, the other one's `process_dict` will also be modified, so they aren't unique. It's one of the subtleties of object oriented programming that can trip you up when you're still new to it.
`process_dict` should be initialized in the `__init__` method and replace `self.T` (with `self.process_dict`).

ottopasuuna commented on ProcessManager.py:
```diff
@@ -52,13 +56,17 @@ def createThread(self):
         """
 
+    def startProcess(self, name):
+        process_dict[name].execute()
```
👍 We should probably check if `name` is in `process_dict` before accessing it.

ottopasuuna replied:
Also, if you could add a [docstring](https://www.python.org/dev/peps/pep-0257/) to this method, that would be great.

chameleon07 replied:
Would the others need a doc string?

ottopasuuna replied:
Yeah, everything need docs. :)

ottopasuuna commented on ProcessManager.py:
```diff
@@ -67,9 +75,9 @@ def stopAllProcesses(self):
-            process.wait()
+        for process in process_dict:
+            process.execute.kill()
+            process.execute.wait()
```
`process.execute` is a function (in python, functions are also objects that can be passed around and accessed like any other variable), so this won't work as it is. If you change the `RoboProcesses.execute()` method to store the Popen object as I explained above, you can just replace `execute` with `popen` (or what ever you choose to call it) and it should work the way you'd expect.

ottopasuuna reviewed (commented):
Great, getting closer!
One non critical thing I noticed is that the name of the `RoboProcesses` class is plural, when it contains only one process. Just for consistancy it might be good to rename it to `RoboProcess`.

climathosphere committed:
Fixed indentation at Line 25.

chameleon07 committed:
Saved Popen object to self.popen
Replaced self.T with self.process_dict
Created loop to check if name has been used

chameleon07 committed:
Added docstring for createProcess

chameleon07 committed:
Added and edited doc strings of methods finished in create-stop-start-proceses branch

ottopasuuna commented on ProcessManager.py:
```diff
@@ -3,50 +3,97 @@
+
         """
+        if type(command) == str:
+            for process in self.process_dict:
```
Instead of looping through all elements in `self.process_dict`, python allows you to do this:
```python
if name in self.process_dict:
    print("Name already taken.")
```

ottopasuuna commented on ProcessManager.py:
```diff
@@ -3,50 +3,97 @@
     This class represents the pocessing module for the robocluster operating system.
     """
-    
+    process_dict = {}
```
This line can be removed now.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -3,50 +3,97 @@
     def __init__(self, **kwargs):
         """Initializes the process manager"""
-        self.T = [] # Thread or process library (It is right now a list but it could be changed to something else other than a list).
+        self.process_dict = {} # Thread or process library (It is right now a list but it could be changed to something else other than a list).
```
This comment should be updated to describe what `process_dict` is storing.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -3,50 +3,97 @@
+
+        for process in self.process_dict:
+            process.execute.kill()
+            process.execute.wait()
```
We'll have to access the popen object like so:
```python
for process in self.process_dict:
    self.process_dict[process].popen.kill()
    ...
```

ottopasuuna replied:
Also could we add a method similar to `startProcess` that stops a single process? It would be usefull later on when restarting processes.

ottopasuuna reviewed (commented):
Basically done, just a few more tweaks and this should be good to get merged for this weekend.

chameleon07 committed:
Corrected code according to review

jpas commented:
I would like to point out that a process manager does a a specific order of execution and needs to handle errors properly so you may want to implement the [contextmanager](https://docs.python.org/3.6/library/stdtypes.html#typecontextmanager) methods to ensure the order of operations is preserved (i.e. `manager.startAllProcesses()` then `manager.stopAllProcesses()`) using the `with` statement.

```py
with ProcessManager() as manager:
    # do stuff with manager
```

ottopasuuna commented:
@jpas I see, thanks for bringing this up. I was thinking the ProcessManager would have a command line and later web interface for starting and stopping processes. Would a context manager still be suitable in this situation? Also, is a contextmanager something that could be added a bit later as part of the process monitering? Just trying to get this merged asap.

ottopasuuna committed:
Made test code in ProcessManager run processes.

Also cleaned up some styling.

ottopasuuna commented:
@chameleon07 I just played around with getting your code to actually run things and it almost works, but needs one important change to do so. I updated the main test code at the bottom of the ProcessManager.py file to start up some processes, and it should be somewhat obvious what the issue is if you run ProcessManager.py. Basically we're storing the command as a string, but `Popen()` takes a list. Storing the command might be more convenient however, so it might be best to change `RoboProcess.execute` to break up the string into a list.

jpas commented:
@ottopasuuna re contextmanager: we could implement it now to simplify the API, but it _could_ be added later. The ProcessManager itself could also have a `cli()` method for our command line interface, there are a bunch of ways to set up the webserver with it as well:

```py
with ProcessManager() as manager:
    manager.start_webserver()
    manager.cli()
```
or
```py
with ProcessManager(webserver=True) as manager:
    manager.follow_log()  # essentially daemon mode
```

@ottopasuuna @chameleon07 re: `Popen(cmd)` I would advise against storing commands as strings due to the potential for improper handling of word splitting.

Using the shell command `./foo "argument with spaces"`.
The equivalent call to Popen would be `Popen(['./foo', 'argument with spaces'])`.
Splitting the string on spaces would yield: `['./foo', '"argument', 'with', 'spaces"']`  which is definitely not what we wanted.

ottopasuuna commented:
Right, somehow I forgot about commands having arguments with spaces despite using them all the time...
Is there a way we could parse commands taking that into account? I'm just thinking it is going to be anoying to write configuration files and interact with a cli if we have to manually separate commands into list syntax. It doesn't matter to me at what stage things are parsed, but it would just be convenient if the rover operator didn't have to do it manually.

After discussing with @jpas, adding contextmanager support should be trivial, just start all the processes in `__enter__()` and shut them all down in `__exit__()`. Then we can be sure that processes will always be terminated when the manager exits.

jpas commented:
@ottopasuuna turns out there's a [builtin library](https://docs.python.org/3.6/library/shlex.html) for shell based lexing!

chameleon07 committed:
Added shlex.split(cmd) to handle word splitting for shell

chameleon07 committed:
Renamed RoboProcesses to RoboProcess and added doctrings to RoboProcess

ottopasuuna committed:
Fixed some minor styling issues

ottopasuuna commented:
This all looks good to me, I just touched up some minor styling issues, and reworded a few docstrings.
@jpas, can you think of any more changes that should be made before merge?

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,150 @@
+        self.cmd = cmd
+        self.popen = None
+
+    def execute(self):
```
We may want to safeguard against, repeated calls of execute.

jpas committed:
Adds contextmanager to ProcessManager

jpas committed:
Renames process_dict to processes

jpas committed:
Creates main method

jpas committed:
Fixes isEmpty, comparing against empty dict is bad.

jpas committed:
Cleans up some docstrings.

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
+        print('Starting {}'.format(name))
+        self.processes[name].execute()
 
     def startAllProcesses(self):
```
I think having this syntax `manager.start()` would be better. We could also use an optional `name` kwarg to start a single process.

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
+        self.processes[name].popen.kill()
+        self.processes[name].popen.wait()
 
     def stopAllProcesses(self):
```
I think having this syntax `manager.stop()` would be better. We could also use an optional `name` kwarg to stop a single process.

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
+        for process in self.processes:
+            self.stopProcess(process)
 
     def status(self, thread):
```
Move into `RoboProcess`

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
+        """Return the status of a thread or process."""
+        raise NotImplementedError()
 
     def verify(self, thread):
```
Move into `RoboProcess`

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
+        """Verify if a thread of process is running properly."""
+        raise NotImplementedError()
 
     def fix(self, thread):
```
Move into `RoboProcess`

jpas commented on ProcessManager.py:
```diff
@@ -1,88 +1,136 @@
 
+        self.processes[name] = RoboProcess(command)
+
+    # TODO: Do we really need this if we have createProcess?
```
I don't think the ProcessManager needs to manage threads. We also shouldn't be running threads within the process manager as they could cause it to exit unexpectedly.

chameleon07 committed:
Changes implemented: method on threads removed, status, verify, and fix are moved, stop and start methods created

ottopasuuna commented on ProcessManager.py:
```diff
@@ -66,47 +77,40 @@ def createProcess(self, name, command):
-        """Verify if a thread of process is running properly."""
-        raise NotImplementedError()
+        if name == []:
+            """Start all managed processes."""
```
Comments that are not docstrings should use the `#` symbol instead of tripple quoted strings. https://www.python.org/dev/peps/pep-0008/#block-comments

ottopasuuna commented on ProcessManager.py:
```diff
@@ -66,47 +77,40 @@ def createProcess(self, name, command):
-    # TODO: Do we really need this if we have createProcess?
-    def createThread(self):
+
+    def start(self,**name):
```
The syntax `**name` takes in a variable number of arguments indexable by keyword:
``` python
def send_msg(**kwargs):
    print(kwargs["test"])

send_msg(test="Hi")
```
ie kwargs is a dictionary.
I think we want more of a bunch of strings to be passed in:
``` python
...
    def start(self, *args):
        for name in args:
             self.processes[name] ... # start the process
...

manager.start("sleeper", "crash", "somethingelse")
```
But first check if the length of `args` is not 0 before looping over it. Or that it exists at all.

ottopasuuna commented on ProcessManager.py:
```diff
@@ -66,47 +77,40 @@ def createProcess(self, name, command):
+            for process in self.processes:
+                self.processes[name].execute()
+        else:
+            for process in name:
```
~If `manager.start()` is called without any parameters, `name` will be empty, so this loop won't run. Loop over `self.processes` instead.~
Edit: Nevermind, I mis-read this somehow. The real problem is 2 and 3 lines below where you use the name list, when it should be the process variable you created in the for loop.

chameleon07 replied:
I should just change the variable so that there won't be any confusion?

jpas replied:
You should change the `[name]` to `[process]`.

ottopasuuna reviewed (changes requested):
All my suggestions for `start()` also applies to `stop()`.
I'd like this merged this Saturday, so as soon as these changes are fixed, I'll merge this into dev. Any other changes can be done in future branches and finalized before a dev is merged with master or the first release is made.

jpas commented on ProcessManager.py:
```diff
@@ -66,47 +77,40 @@ def createProcess(self, name, command):
-    def verify(self, thread):
-        """Verify if a thread of process is running properly."""
-        raise NotImplementedError()
+        if name == []:
```
> For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
https://www.python.org/dev/peps/pep-0008/#programming-recommendations

jpas commented on ProcessManager.py:
```diff
@@ -66,47 +77,40 @@ def createProcess(self, name, command):
+                self.processes[name].execute()
+
+    def stop(self,**name):
+        if name == []:
```
> For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
https://www.python.org/dev/peps/pep-0008/#programming-recommendations

chameleon07 committed:
Replaced process of checking if name is empty and corrected name variable with process in start and stop methods

jpas committed:
Prevent sleeper from complaining about KeyboardInterrupt

jpas committed:
Fixes docstrings

jpas committed:
Cleans up start/stop implementation
